apps/dcos_dns/src/dcos_dns_handler.erl

```diff
@@ -177,90 +177,106 @@ resolve_loop(Begin, Workers, Fun, FailMsg, Zone) ->
 -spec(start_worker(protocol(), upstream(), binary()) -> pid()).
 start_worker(Protocol, Upstream, Request) ->
     Pid = self(),
+    Begin = erlang:monotonic_time(),
+    BeginMilliseconds = erlang:monotonic_time(millisecond),
     proc_lib:spawn(fun () ->
-        init_worker(Protocol, Pid, Upstream, Request)
+        UpstreamAddress = upstream_to_binary(Upstream),
+        case init_worker(Protocol, Pid, Upstream, Request) of
+            {ok, Response} ->
+                Pid ! {reply, self(), Response},
+                report_latency([?MODULE, Upstream, latency], BeginMilliseconds),
+                prometheus_histogram:observe(
+                    dns, worker_requests_duration_seconds,
+                    [Protocol, UpstreamAddress],
+                    erlang:monotonic_time() - Begin);
+            {error, Reason} ->
+                lager:warning(
+                    "DNS worker [~p] ~p failed with ~p",
+                    [Protocol, UpstreamAddress, Reason]),
+                prometheus_counter:inc(
+                    dns, worker_failures_total,
+                    [Protocol, UpstreamAddress, Reason], 1)
+        end
     end).
 
--spec(init_worker(protocol(), pid(), upstream(), binary()) -> ok).
-init_worker(udp, Pid, Upstreams, Request) ->
-    udp_worker(Pid, Upstreams, Request);
-init_worker(tcp, Pid, Upstreams, Request) ->
-    tcp_worker(Pid, Upstreams, Request).
+-spec(init_worker(protocol(), pid(), upstream(), binary()) ->
```

Comment:
There were 3 functions that tail called each other. At the moment, `init_worker/4` does much more than just worker initialization. Could you please rename it to `worker_resolve(…)` or something like this?